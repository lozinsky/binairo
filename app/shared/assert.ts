class AssertionError extends Error {
  constructor(message?: string) {
    super(message);

    this.name = 'AssertionError';
  }
}

export function assert(condition: boolean, message?: string): asserts condition {
  if (!condition) {
    throw new AssertionError(message ?? 'Expected condition to be true');
  }
}

export function unreachable(message?: string): never {
  throw new AssertionError(message ?? 'Reached unreachable code');
}
