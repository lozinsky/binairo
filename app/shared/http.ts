export function getErrorResponse(error: unknown): unknown {
  if (error instanceof Error && error.name === 'AssertionError') {
    return new Response(error.message, { status: 400 });
  }

  return error;
}
