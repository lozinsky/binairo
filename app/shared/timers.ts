export interface Abortable {
  signal?: AbortSignal;
}

export async function setTimeout<T = void>(delay?: number, value?: T, options?: Abortable) {
  let resolve!: (value: T) => void;
  let reject!: (reason: unknown) => void;

  const args = ([] as T[]).concat(value ?? []);
  const signal = options?.signal;
  const promise = new Promise<T>((resolveCallback, rejectCallback) => {
    resolve = resolveCallback;
    reject = rejectCallback;
  });

  if (signal === undefined) {
    globalThis.setTimeout(resolve, delay, ...args);
    return await promise;
  }

  if (signal.aborted) {
    reject(signal.reason);
    return await promise;
  }

  const handleAbort = () => {
    globalThis.clearTimeout(handle);
    reject(signal.reason);
  };

  signal.addEventListener('abort', handleAbort);

  const handle = globalThis.setTimeout(resolve, delay, ...args);

  try {
    return await promise;
  } finally {
    signal.removeEventListener('abort', handleAbort);
  }
}
