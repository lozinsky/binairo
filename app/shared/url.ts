export function resolvePathname(pathname: string) {
  return pathname.replace(/^\//, import.meta.env.BASE_URL);
}
