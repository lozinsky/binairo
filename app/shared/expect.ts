import { assert } from '~/shared/assert';

export function expectToSatisfy<T>(value: unknown, predicate: (value: unknown) => value is T, message?: string) {
  assert(predicate(value), message ?? 'Expected value to satisfy the predicate');

  return value;
}

export function expectToBeDefined<T>(value: T, message?: string) {
  assert(value != null, message ?? 'Expected value to be defined');

  return value;
}

export function expectNotToBeNaN(value: number, message?: string) {
  assert(!isNaN(value), message ?? 'Expected value not to be "NaN"');

  return value;
}
