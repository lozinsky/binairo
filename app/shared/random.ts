import { type RandomGenerator, uniformIntDistribution, xoroshiro128plus } from 'pure-rand';

import { expectToBeDefined } from '~/shared/expect';

export class Random {
  #generator: RandomGenerator;
  #seed: number;

  constructor(seed: number) {
    this.#generator = xoroshiro128plus(seed);
    this.#seed = seed;
  }

  static create() {
    return new this(Date.now() ^ (Math.random() * 0x100000000));
  }

  static stable() {
    return new this(0);
  }

  next(from: number, to: number) {
    const [value, generator] = uniformIntDistribution(from, to, this.#generator);

    this.#generator = generator;

    return value;
  }

  get seed() {
    return this.#seed;
  }
}

export function shuffle<T>(target: readonly T[], random: Random) {
  const items = [...target];

  for (let prevIndex = 0; prevIndex < items.length; prevIndex++) {
    const nextIndex = random.next(prevIndex, items.length - 1);
    const prevItem = expectToBeDefined(items[prevIndex]);
    const nextItem = expectToBeDefined(items[nextIndex]);

    items[prevIndex] = nextItem;
    items[nextIndex] = prevItem;
  }

  return items;
}

export function sample<T>(target: readonly T[], random: Random) {
  return expectToBeDefined(target[random.next(0, target.length - 1)]);
}
