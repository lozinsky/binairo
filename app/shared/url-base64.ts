import { expectToBeDefined } from '~/shared/expect';

const PADDING_REGEX = /[.=]{1,2}$/;

const URL_UNSAFE_CHAR_REGEX = /[+/]/g;

const URL_SAFE_CHAR_REGEX = /[-_]/g;

const URL_UNSAFE_CHAR_BY_URL_SAFE_CHAR: Record<string, string> = {
  '-': '+',
  _: '/',
};

const URL_SAFE_CHAR_BY_URL_UNSAFE_CHAR: Record<string, string> = {
  '+': '-',
  '/': '_',
};

export function decodeURLBase64(value: string) {
  return atob(
    value.replaceAll(URL_SAFE_CHAR_REGEX, (substring) =>
      expectToBeDefined(URL_UNSAFE_CHAR_BY_URL_SAFE_CHAR[substring]),
    ),
  );
}

export function encodeURLBase64(value: string) {
  return btoa(value)
    .replaceAll(URL_UNSAFE_CHAR_REGEX, (substring) => expectToBeDefined(URL_SAFE_CHAR_BY_URL_UNSAFE_CHAR[substring]))
    .replace(PADDING_REGEX, '');
}
