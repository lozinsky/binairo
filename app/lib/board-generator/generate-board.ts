import { Board, BoardCell, BoardCellState } from '~/lib/board';
import {
  selectFilledImbalancedLine,
  selectLineThatIsEqualToOthers,
  selectThreeOrMoreIdenticalSequentialCells,
} from '~/lib/board-selectors';
import { MatrixSelection } from '~/lib/matrix';
import { expectToBeDefined } from '~/shared/expect';
import { type Random, shuffle } from '~/shared/random';
import { type Abortable, setTimeout } from '~/shared/timers';

import { generateBoardLines } from './generate-board-lines';

function isValidBoard(target: Board) {
  if (selectFilledImbalancedLine(target) !== undefined) {
    return false;
  }

  if (selectLineThatIsEqualToOthers(target) !== undefined) {
    return false;
  }

  if (selectThreeOrMoreIdenticalSequentialCells(target) !== undefined) {
    return false;
  }

  return true;
}

export async function generateBoard(size: number, progress: number, random: Random, options?: Abortable) {
  const lines = generateBoardLines(size);
  const combinations = shuffle(lines, random);

  let attempt = 0;
  let board = Board.blank(size);
  let index = 0;

  do {
    const line = expectToBeDefined(combinations.shift());
    const target = board.replace(index, () => line);

    attempt++;

    if (isValidBoard(target)) {
      board = target;
      index++;
      attempt = 0;
    } else {
      combinations.push(line);

      if (attempt >= combinations.length) {
        for (const line of board) {
          if (lines.includes(line)) {
            combinations.push(line);
          }
        }

        board = Board.blank(size);
        index = 0;
        attempt = 0;

        await setTimeout(undefined, undefined, options);
      }
    }
  } while (index < size);

  do {
    const selection = new MatrixSelection([
      {
        x: random.next(0, size - 1),
        y: random.next(0, size - 1),
      },
    ]);

    board = board.replaceBy(selection, () => BoardCell.create(BoardCellState.E));
  } while (board.progress > progress);

  return board;
}
