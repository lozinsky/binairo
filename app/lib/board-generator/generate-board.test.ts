/* eslint-disable @typescript-eslint/no-floating-promises */

import { expect, test } from 'vitest';

import { Random } from '~/shared/random';

import { generateBoard } from './generate-board';

test.each([
  [4, 0.6],
  [6, 0.6],
  [8, 0.6],
  [10, 0.6],
  [12, 0.6],
])('generates board', (size, progress) => {
  expect(generateBoard(size, progress, Random.stable())).resolves.toMatchSnapshot();
});
