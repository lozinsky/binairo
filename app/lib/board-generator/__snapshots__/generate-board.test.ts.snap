// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generates board 1`] = `
[⬜] {🟦} [⬜] {🟦}
[⬜] [⬜] {🟦} [⬜]
{🟦} [⬜] {🟥} [⬜]
{🟦} {🟥} {🟦} {🟥}
`;

exports[`generates board 2`] = `
{🟦} {🟦} {🟥} {🟥} [⬜] {🟥}
{🟥} {🟥} {🟦} {🟥} [⬜] [⬜]
[⬜] [⬜] [⬜] {🟦} {🟥} [⬜]
[⬜] [⬜] {🟥} {🟦} {🟥} [⬜]
{🟥} [⬜] {🟦} {🟥} [⬜] [⬜]
{🟥} {🟥} [⬜] [⬜] {🟥} {🟦}
`;

exports[`generates board 3`] = `
{🟥} [⬜] {🟥} {🟦} [⬜] {🟥} {🟦} {🟥}
{🟦} {🟥} {🟦} [⬜] [⬜] {🟦} [⬜] {🟥}
[⬜] {🟥} {🟥} {🟦} [⬜] {🟥} {🟥} {🟦}
{🟥} {🟦} [⬜] {🟥} {🟥} [⬜] [⬜] {🟥}
[⬜] [⬜] {🟥} [⬜] [⬜] [⬜] {🟥} {🟦}
[⬜] [⬜] [⬜] {🟦} {🟥} {🟥} {🟦} [⬜]
{🟦} [⬜] {🟥} {🟦} {🟥} [⬜] {🟥} [⬜]
[⬜] {🟦} [⬜] [⬜] {🟦} {🟥} {🟥} [⬜]
`;

exports[`generates board 4`] = `
[⬜] {🟥} [⬜] [⬜] [⬜] {🟥} [⬜] {🟥} {🟦} [⬜]
{🟥} [⬜] {🟦} {🟥} [⬜] [⬜] [⬜] {🟦} {🟥} [⬜]
{🟥} {🟦} {🟥} [⬜] {🟦} {🟥} {🟦} {🟥} [⬜] [⬜]
{🟦} {🟥} {🟦} {🟥} {🟦} {🟦} {🟥} [⬜] {🟥} {🟥}
[⬜] [⬜] {🟥} {🟦} {🟥} [⬜] {🟥} {🟥} [⬜] {🟦}
{🟥} [⬜] {🟦} {🟥} {🟦} {🟥} [⬜] {🟦} {🟥} [⬜]
{🟦} [⬜] [⬜] {🟦} {🟥} {🟦} {🟥} [⬜] [⬜] {🟥}
{🟦} {🟥} {🟦} {🟥} [⬜] [⬜] [⬜] {🟥} [⬜] {🟦}
{🟥} [⬜] {🟥} [⬜] [⬜] [⬜] {🟦} [⬜] {🟥} {🟥}
[⬜] {🟥} {🟦} {🟥} [⬜] {🟦} [⬜] {🟦} [⬜] [⬜]
`;

exports[`generates board 5`] = `
{🟥} [⬜] [⬜] {🟥} [⬜] {🟥} {🟦} [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] {🟥} {🟥} {🟦} [⬜] {🟦} [⬜] [⬜] [⬜] {🟥} [⬜] [⬜]
{🟦} {🟥} [⬜] [⬜] {🟦} {🟥} {🟦} {🟥} [⬜] {🟥} [⬜] [⬜]
{🟥} [⬜] {🟥} {🟦} [⬜] {🟦} [⬜] [⬜] [⬜] {🟦} [⬜] [⬜]
{🟥} {🟦} {🟥} {🟥} {🟦} [⬜] {🟥} {🟦} [⬜] {🟥} {🟥} {🟦}
[⬜] [⬜] {🟦} {🟥} {🟦} [⬜] {🟥} {🟦} {🟥} {🟦} [⬜] {🟦}
[⬜] {🟦} {🟥} [⬜] {🟥} [⬜] {🟦} {🟥} {🟥} {🟦} {🟦} {🟥}
[⬜] {🟥} {🟦} [⬜] {🟥} {🟦} [⬜] {🟥} [⬜] [⬜] [⬜] [⬜]
{🟦} {🟦} {🟥} {🟥} {🟦} {🟥} {🟥} [⬜] {🟦} [⬜] [⬜] [⬜]
[⬜] {🟥} [⬜] [⬜] {🟥} {🟥} {🟦} [⬜] {🟥} [⬜] {🟦} [⬜]
{🟥} {🟥} {🟦} {🟦} {🟥} {🟦} {🟥} {🟥} [⬜] {🟦} [⬜] [⬜]
{🟥} {🟦} [⬜] {🟥} {🟦} {🟦} {🟥} {🟦} [⬜] {🟥} {🟦} {🟥}
`;
